### PR TITLE
[Feature] [Algorithm] Add FQL Loss

### PR DESCRIPTION
**Pull request by neuriv**
## Description

Add FQLLoss to train the flow policy on replay actions, distill its samples into a one-step actor and improve that actor using critic values. Train the critics with target values while supporting TorchRL’s loss keys, masks, sample weights and target updates.

Part of #4433.

<details>
<summary>Testing</summary>

Ubuntu 22.04, A100 40 GB, Python 3.11.16, PyTorch 2.14.0+cu126 and TensorDict 0.14.2.

Run from the `feat/fql` worktree. Additional test scripts were kept outside the repository.

```bash
root=/home/ubuntu/fql-validation-20260919
suite="$root/tests"
evidence="$root/evidence"
export PATH="$root/rl/.venv/bin:$PATH"
export PYTHONPATH="$PWD" FQL_TEST_DEVICE=cpu,cuda COMPOSITE_LP_AGGREGATE=0
export CUDA_VISIBLE_DEVICES=0 MUJOCO_GL=egl OMP_NUM_THREADS=1
export HYDRA_FULL_ERROR=1 WANDB_MODE=disabled WANDB_SILENT=true

pre-commit run --from-ref b4caa999e7619cc416174d682725749c2079f0c0 --to-ref HEAD
python -m pytest -q "$suite/test_flow.py" "$suite/test_fql.py" test/objectives/test_loss_module.py

for seed in 17 0 1; do
    python "$suite/check_fql.py" --device cuda --seed "$seed" --output "$evidence/fql/seed-$seed"
    python "$suite/verify_fql_evidence.py" "$evidence/fql/seed-$seed" --device cuda
done
```

362 tests passed. Pre-commit passed.

| Seed | Initial action MSE | FQL action MSE | Behavior-only action MSE | FQL critic MAE | Frozen-target critic MAE |
| --- | ---: | ---: | ---: | ---: | ---: |
| 0 | 0.397491 | 0.010882 | 1.065319 | 0.003148 | 1.044985 |
| 1 | 0.279830 | 0.010165 | 1.021522 | 0.004088 | 0.962058 |
| 17 | 0.425254 | 0.007791 | 1.052302 | 0.004853 | 0.980475 |

All three seeds passed. Reloaded checkpoints reproduced the measurements and passed independent Bellman-value, target-weight and optimizer-step checks.

</details>

### CODE REVIEW

**Review by vmoens — COMMENTED** [review]
Needs rebase
We should add some tests, in particular compile compatibility. Further steps are also perhaps a benchmark and a Trainer class + configs (in follow-up PRs). Would love to see a bunch of training curves.

File: torchrl/objectives/fql.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        )
+
+    def flow_loss(self, tensordict):
+        action = tensordict.get(self.tensor_keys.action)
+        observation = tensordict.get(self.tensor_keys.observation)
+        noise = torch.randn_like(action)
+        time = torch.rand_like(action[..., :1])
+        interpolated = (1 - time) * noise + time * action
+        with self.flow_policy_params.to_module(self.flow_policy):
+            velocity = self.flow_policy.velocity(observation, interpolated, time)
```
Comment (theap06) [inline torchrl/objectives/fql.py:134]: this reaches into `flow_policy.velocity` and `actor_network(obs, noise, clamp=False)` with raw tensors, so the policies' in_keys are ignored and nested/multi-key obs or an encoder cant be routed. can the loss go through the td interface like the other losses?

**Review by theap06 — APPROVED** [review]
LGTM! Thanks. I'll take a look at the remaining prs in this "stack"
